### ISSUE TITLE
Suggestion: disallow implicit return of undefined

### ISSUE THREAD

**Issue by neilgalarneau** [Microsoft/TypeScript#2316]
It would be useful to be able to disallow an Explicitly Typed function/method from implicitly returning the value undefined.

One example function that would be detected and disallowed was given by @rolyp in #2176. In that example, running off the end of the function causes undefined to be returned implicitly.

Another example that should be caught is:
```typescript
function u(): number {
  return;
}
```

A flag for enabling this option could be named --noImplictReturnUndefined.

This suggestion fits in with the meta-bug in #274 and maybe #2176.

**Comment by danquirk** [comment]
Definitely fits well within either of those two larger options.

**Comment by RyanCavanaugh** [comment]
We should use this as an opportunity to look at code patterns in the wild -- how many JavaScript|TypeScript functions "intentionally" use an expressioned return and unexpressioned return in the same function body, especially in those with return type annotations.

**Comment by neilgalarneau** [comment]
I understand that TypeScript's goal is to be a super set of JavaScript, so legal JavaScript is legal TypeScript.

However, if I add types to my code, it is no longer JavaScript, and I would like TypeScript to help me fulfill the contract that the type usage represents.

If I declare a return type on my function and that function ends with "return;", I would like TypeScript to see that as an inconsistency & point that out to me.

**Comment by mhegazy** [comment]
`--noImplicitReturns` will flag missing a return statement. the main issue today is that undefined is a valid value for any type, and `return;` is equivalent to `return undefined`; i would say this falls under the bucket of https://github.com/Microsoft/TypeScript/issues/185
